common/OpTestSystem.py

```diff
@@ -222,91 +385,182 @@ def detect_target(self, target_state, reboot):
-                  previous_before = 'emptyagain'
-                else:
-                  console = self.console.connect()
-                  if buffer_kicker:
-                    console.sendline("\r")
-                  previous_before = 'emptyagain'
+        timeout_count = 1
+        while (x <= kwargs['loop_max']):
+            sys_console = self.console.get_console() # preemptive in case EOF came
+            r = sys_console.expect(expect_seq, kwargs['timeout'])
+            # if we have a stale buffer and we are still timing out
+            if (previous_before == sys_console.before) and ((r + 1) in range(len(base_seq))):
+              timeout_count += 1
+              # only attempt reconnect if we've timed out per threshold
+              if (timeout_count % kwargs['threshold'] == 0):
+                if kwargs['reconnect']:
+                  reconnect_count += 1
+                  try:
+                    sys_console = self.console.connect()
+                  except Exception as e:
+                    print e
+                  if kwargs['refresh']:
+                    sys_console.sendcontrol('l')
+                  if kwargs['buffer_kicker']:
+                    sys_console.sendline("\r")
+                    sys_console.expect("\n")
+                previous_before = 'emptyagain'
             else:
-              previous_before = console.before
-            if (rc == 0):
-              break;
+              previous_before = sys_console.before
+              timeout_count = 1
+
+            working_r = self.check_it(my_r=r, check_base_seq=base_seq, check_expect_seq=expect_seq, check_expect_dict=kwargs['expect_dict'])
+            # if we found a hit on the callers string return it, otherwise keep looking
+            if working_r != -1:
+              return working_r, reconnect_count
             else:
               x += 1
-            if (x >= loop_max):
-              raise Exception('Waiting for "{}" did not succeed, check the loop_max if needing to wait longer'
-                              ' (number of reconnect attempts were {})'.format(my_string, reconnect_count))
+              print ("\n *** WaitForIt CURRENT STATE \"{:02}\" TARGET STATE \"{:02}\"\n"
+                      " *** WaitForIt working on transition\n"
+                      " *** Current loop iteration \"{:02}\"             - Reconnect attempts \"{:02}\" - loop_max \"{:02}\"\n"
+                      " *** WaitForIt timeout interval \"{:02}\" seconds - Stale buffer check every \"{:02}\" times\n"
+                      " *** WaitForIt variables \"{}\"\n"
+                      " *** WaitForIt Refresh=\"{}\" Buffer Kicker=\"{}\" - Kill Cord=\"{:02}\"\n".format(self.state, self.target_state,
+                      x, reconnect_count, kwargs['loop_max'], kwargs['timeout'], kwargs['threshold'],
+                      sorted(kwargs['expect_dict'].keys()), kwargs['refresh'], kwargs['buffer_kicker'], self.kill_cord))
```

Comment:
Atleast comment this statement for now, until your logging infrastructure PR merges.

Reply:
> comment this statement for now

Fixed